src/app/shared/navigation-focus/navigation-focus.spec.ts

```diff
@@ -0,0 +1,103 @@
+import {Component, NgZone} from '@angular/core';
+import {ComponentFixture, fakeAsync, inject, TestBed, tick} from '@angular/core/testing';
+import {Router} from '@angular/router';
+import {RouterTestingModule} from '@angular/router/testing';
+import {NavigationFocusService} from './navigation-focus.service';
+
+describe('Navigation focus service', () => {
+  let navigationFocusService: NavigationFocusService;
+  let router: Router;
+  let zone: NgZone;
+  let fixture: ComponentFixture<NavigationFocusTest>;
+
+  const navigate = (url: string) => {
+    zone.run(() => router.navigateByUrl(url));
+    tick(100);
+  };
+
+  beforeEach(() => {
+      TestBed.configureTestingModule({
+        imports: [RouterTestingModule.withRoutes([
+          {path: '', component: RouteTest},
+          {path: 'cdk', component: RouteTest},
+          {path: 'guides', component: RouteTest}
+        ])],
+        providers: [NavigationFocusService],
+        declarations: [NavigationFocusTest, RouteTest],
+      });
+      fixture = TestBed.createComponent(NavigationFocusTest);
+    }
+  );
+
+  beforeEach(inject(
+    [NgZone, Router, NavigationFocusService],
+    (_zone: NgZone, _router: Router, nfs: NavigationFocusService) => {
+      zone = _zone;
+      router = _router;
+      navigationFocusService = nfs;
+    }));
+
+  it('should set skip link href', () => {
+    const target1 = fixture.nativeElement.querySelector('#target1');
+    const target2 = fixture.nativeElement.querySelector('#target2');
+
+    navigationFocusService.requestSkipLinkFocus(target1);
+    navigationFocusService.requestSkipLinkFocus(target2);
+
+    expect(navigationFocusService.getSkipLinkHref()).toEqual('/#target2');
+
+    navigationFocusService.relinquishSkipLinkFocusOnDestroy(target2);
+
+    expect(navigationFocusService.getSkipLinkHref()).toEqual('/#target1');
+  });
+
+  it('should be within component view', () => {
+    const previousUrl = '/components/autocomplete/overview';
+    const newUrl = '/components/autocomplete/overview#simple-autocomplete';
+    expect(navigationFocusService.isNavigationWithinComponentView(previousUrl, newUrl)).toBeTrue();
+  });
+
+  it('should not be within component view', () => {
+    const previousUrl = '/cdk/clipboard/overview';
+    const newUrl = '/cdk/categories';
+    expect(navigationFocusService.isNavigationWithinComponentView(previousUrl, newUrl)).toBeFalse();
+  });
+
+  it('should focus on component then relinquish focus', fakeAsync(() => {
```

Comment:
Excellent test. It feels really satisfying to see tests demonstrate expected behavior so clearly. Can you also add one that exercises the path where you navigate to a hash target within the page? And one that ensures focus is still applied to elements that didn't have an id (the path where it adds one using the uid).